app/upgrades/v17/convert.go

```diff
@@ -0,0 +1,145 @@
+)
+
+// ConvertERC20Coins converts Native IBC coins from their ERC20 representation
+// to the native representation. This also includes the withdrawal of WEVMOS tokens
+// to EVMOS native tokens.
+func ConvertERC20Coins(
+	ctx sdk.Context,
+	logger log.Logger,
+	accountKeeper authkeeper.AccountKeeper,
+	bankKeeper bankkeeper.Keeper,
+	erc20Keeper erc20keeper.Keeper,
+	wrappedAddr common.Address,
+) error {
+	// iterate over all the accounts and convert the tokens to native coins
+	accountKeeper.IterateAccounts(ctx, func(account authtypes.AccountI) (stop bool) {
+		cosmosAddress := account.GetAddress()
+		ethAddress := common.BytesToAddress(cosmosAddress.Bytes())
+		ethHexAddr := ethAddress.String()
+
+		balance, res, err := WithdrawWEVMOS(ctx, ethAddress, wrappedAddr, erc20Keeper)
+
+		var bs string // NOTE: this is necessary so that there is no panic if balance is nil when logging
+		if balance != nil {
+			bs = balance.String()
+		}
+
+		if err != nil {
+			logger.Error(
+				"failed to withdraw WEVMOS",
+				"account", ethHexAddr,
+				"balance", bs,
+				"error", err.Error(),
+			)
+		} else if res != nil && res.VmError != "" {
+			logger.Error(
+				"withdraw WEVMOS reverted",
+				"account", ethHexAddr,
+				"balance", bs,
+				"vm-error", res.VmError,
+			)
+		}
+
+		erc20Keeper.IterateTokenPairs(ctx, func(tokenPair erc20types.TokenPair) bool {
+			// NOTE: here we check if the token pair contains an IBC coin. For now we only want to convert those.
+			if !tokenPair.IsNativeCoin() {
+				return false
+			}
+
+			contract := tokenPair.GetERC20Contract()
+			if err := ConvertERC20Token(ctx, ethAddress, contract, cosmosAddress, erc20Keeper); err != nil {
+				logger.Error(
+					"failed to convert ERC20 to native Coin",
+					"account", ethHexAddr,
+					"erc20", contract.String(),
+					"balance", balance.String(),
+					"error", err.Error(),
+				)
+			}
+
+			return false
+		})
+
+		return false
+	})
+
+	// NOTE: if there are tokens left in the ERC-20 module account
+	// we return an error because this implies that the migration of native
+	// coins to ERC-20 tokens was not fully completed.
+	erc20ModuleAccountAddress := authtypes.NewModuleAddress(erc20types.ModuleName)
+	balances := bankKeeper.GetAllBalances(ctx, erc20ModuleAccountAddress)
+	if !balances.IsZero() {
+		return fmt.Errorf("there are still tokens in the erc-20 module account: %s", balances.String())
+	}
+
+	return nil
+}
+
+// ConvertERC20Token converts the given ERC20 token to the native representation.
+func ConvertERC20Token(
+	ctx sdk.Context,
+	from, contract common.Address,
+	receiver sdk.AccAddress,
+	erc20Keeper erc20keeper.Keeper,
+) error {
+	balance := erc20Keeper.BalanceOf(ctx, contracts.ERC20MinterBurnerDecimalsContract.ABI, contract, from)
+	if balance == nil {
+		return fmt.Errorf("failed to get ERC20 balance (contract %q) for %s", contract.String(), from.String())
+	}
+
+	if balance.Sign() <= 0 {
+		return nil
```

Comment:
Shouldn't we return an error here? Is this to prevent errors in the upgrade if someone intentionally create an erc20 token with less then 0 balance in some account?

Reply:
We are only converting native Cosmos coins, where the Evmos binary itself creates the corresponding smart contracts and we control the ABI. Also the `<=` check really is only a safety measure, since the balance on ERC-20 contracts is stored as a `uint` so if anything it can only be 0.

The reason we are not returning an error is that it's expected that not every account has an existing balance for every ERC-20 representation on Evmos (if that was the case Evmos DeFi would be BUZZING 😆). So if an account has no balance for the given token pair we return `nil` and continue with the next token pair.